From: Ben Hutchings <ben@decadent.org.uk>
Date: Mon, 04 Jan 2016 19:34:57 +0000
Subject: syscalls: Add accept() wrapper for accept4()
Bug-Debian: https://bugs.debian.org/809423

i386 now has direct socket syscalls, but they don't include accept().
Define accept() as a wrapper for the accept4() syscall if available.

Signed-off-by: Ben Hutchings <ben@decadent.org.uk>
---
--- a/usr/klibc/socketcalls.pl
+++ b/usr/klibc/socketcalls.pl
@@ -47,7 +47,23 @@ while ( defined($line = <FILE>) ) {
 
 	print OUT "#include \"socketcommon.h\"\n";
 	print OUT "\n";
-	print OUT "#if _KLIBC_SYS_SOCKETCALL || !defined(__NR_${name})\n\n";
+
+	if ($name eq 'accept') {
+	    print OUT "#if !_KLIBC_SYS_SOCKETCALL && defined(__NR_accept4)\n\n";
+
+	    print OUT "extern int accept4(int, struct sockaddr *, socklen_t *, int);\n\n";
+
+	    print OUT "int accept(int a0, struct sockaddr *a1, socklen_t *a2)\n";
+	    print OUT "{\n";
+	    print OUT "    return accept4(a0, a1, a2, 0);\n";
+	    print OUT "}\n\n";
+
+	    print OUT "#elif ";
+	} else {
+	    print OUT "#if ";
+	}
+
+	print OUT "_KLIBC_SYS_SOCKETCALL || !defined(__NR_${name})\n\n";
 
 	print OUT "extern long __socketcall(int, const unsigned long *);\n\n";
 
--- a/usr/klibc/SOCKETCALLS.def
+++ b/usr/klibc/SOCKETCALLS.def
@@ -9,6 +9,7 @@
 <?> int connect(int, const struct sockaddr *, socklen_t);
 <?> int listen(int, int);
 <?> int accept(int, struct sockaddr *, socklen_t *);
+<?> int accept4(int, struct sockaddr *, socklen_t *, int);
 <?> int getsockname(int, struct sockaddr *, socklen_t *);
 <?> int getpeername(int, struct sockaddr *, socklen_t *);
 <?> int socketpair(int, int, int, int *);
